examples/expect-extend/toBeWithinRange.ts

```diff
@@ -0,0 +1,43 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import {expect} from '@jest/globals';
+import type {RawMatcherFn} from 'expect';
+
+const toBeWithinRange: RawMatcherFn = (
+  actual: number,
+  floor: number,
+  ceiling: number,
+) => {
+  const pass = actual >= floor && actual <= ceiling;
+  if (pass) {
+    return {
+      message: () =>
+        `expected ${actual} not to be within range ${floor} - ${ceiling}`,
+      pass: true,
+    };
+  } else {
+    return {
+      message: () =>
+        `expected ${actual} to be within range ${floor} - ${ceiling}`,
+      pass: false,
+    };
+  }
+};
+
+expect.extend({
+  toBeWithinRange,
+});
+
+declare module 'expect' {
+  interface AsymmetricMatchers {
+    toBeWithinRange(a: number, b: number): void;
+  }
+  interface Matchers<R> {
+    toBeWithinRange(a: number, b: number): R;
+  }
+}
```

Comment:
there's no way we can avoid the duplication here, right?

Reply:
Wrestled with this. No luck. Alternative idea: what if `expect.extend()` would return `Expect` extended with types of custom matchers? Right now `expect.extend()` returns `void`. The proposed return value would be only useful for typings. No need to extend, augment, etc.

```ts
import {expect as jestExpect} from '@jest/globals';
import customMatcher from 'customMatcher';

const expect = jestExpect.extend({
  customMatcher,
});

// here `expect.customMatcher()` is fully typed
```

Reply:
Drafty, but working prototype (in this case, `toBeWithinRange` takes two arguments and both are inferred):

<img width="757" alt="Screenshot 2022-02-11 at 20 50 11" src="https://user-images.githubusercontent.com/72159681/153651813-c7d31591-a408-426b-9dbe-947ada6b43b3.png">

Reply:
I like it! 😃 We need "old" way to continue working of course, but doing it this way seems smoother